scapy/layers/ipsec.py

```diff
@@ -127,8 +129,38 @@ class ESP(Packet):
 
 bind_layers(IP, ESP, proto=socket.IPPROTO_ESP)
 bind_layers(IPv6, ESP, nh=socket.IPPROTO_ESP)
-bind_layers(UDP, ESP, dport=4500)  # NAT-Traversal encapsulation
-bind_layers(UDP, ESP, sport=4500)  # NAT-Traversal encapsulation
+
+
+class NAT_T(Packet):
+    """
+    UDP Encapsulation of IPsec ESP Packets [RFC3948] (for NAT-Traversal)
+    """
+    name = 'NAT_T'
+
+    __slots__ = [
+        "_esp_spi"
+    ]
+
+    fields_desc = [
+        ConditionalField(XIntField("non_esp", 0x0),
+                         lambda pkt: isinstance(pkt.payload, ISAKMP_class) or pkt._esp_spi == 0x00)  # noqa: E501
+    ]
+
+    def __init__(self, _pkt=b"", *args, **kwargs):
+        if isinstance(_pkt, bytes) and len(_pkt) >= 4:
+            self._esp_spi = struct.unpack("!I", _pkt[0:4])[0]
+        else:
+            self._esp_spi = None
+        super(NAT_T, self).__init__(_pkt=_pkt, *args, **kwargs)
```

Comment:
It would be cleaner to move this into a `pre_dissect` function

Reply:
BTW: I was wandering what to do with `self._esp_spi` after first-use: should it be cleared immediately? If not, how should I handle the case when the payload is replaced programmatically?

Reply:
_(I'm asking because it is used in the `ConditionalField`)_.

Reply:
Answering my own question: clear it in the `post_dissect` hook?

Reply:
addressed in 98be384847b34ad42ee4a9f779b3a24a6500c265